### ISSUE TITLE
Invalid memory access when using pointers for iterator arguments

### ISSUE THREAD

**Issue by SlayerShadow** [crystal-lang/crystal#10835]
Probably similar to #8636, #8482.

```crystal
class User
  def initialize
    @a = 5
  end
end

def show( pointers : Array( Pointer( Void* ) ) )
  p pointers.first.as( Pointer( User ) ).value
  p pointers.last.as( Pointer( StaticArray( User?, 2 ) ) ).value
end

user = User.new
array = StaticArray[ user, nil ]

raw_array = [
  Pointer( Void* ).new( pointerof( user ).address ),
  Pointer( Void* ).new( pointerof( array ).address )
]

map_array = [ user, array ].map{|value|
  # The problem is here
  Pointer( Void* ).new pointerof( value ).address
}

p raw_array, map_array

show raw_array
show map_array
```

Produces the output:

```
[Pointer(Pointer(Void))@0x7ffd602b1ec0, Pointer(Pointer(Void))@0x7ffd602b1ea0]
[Pointer(Pointer(Void))@0x7ffd602b1de8, Pointer(Pointer(Void))@0x7ffd602b1de8]

#<User:0x7fdc009f5eb0 @a=5>
StaticArray[#<User:0x7fdc009f5eb0 @a=5>, nil]

#<User:0x18a @a=Invalid memory access (signal 11) at address 0x18e
[0x5598235eaaa6] *Exception::CallStack::print_backtrace:(Int32 | Nil) +118
[0x5598235ddf8a] ~procProc(Int32, Pointer(LibC::SiginfoT), Pointer(Void), Nil) +330
[0x7fdc00e6bc89] ???
```
( https://carc.in/#/r/bdmk )

Please note that `map_array` pointers has single address `0x7ffd602b1de8`.

It seems that Union is important here: change array elements both to `user` or `array` makes pointers different (what?) but works correctly (please don't forget to patch `def show` if you'd like to run it):

```crystal
raw_array = [
  Pointer( Void* ).new( pointerof( user ).address ),
  Pointer( Void* ).new( pointerof( user ).address )
]

map_array = [ user, user ].map{ |value| Pointer( Void* ).new pointerof( value ).address }
```

Output:

```
[Pointer(Pointer(Void))@0x7ffcb35dee90, Pointer(Pointer(Void))@0x7ffcb35dee90]
[Pointer(Pointer(Void))@0x7ffcb35dede8, Pointer(Pointer(Void))@0x7ffcb35dede8]

#<User:0x7f7af20b1eb0 @a=5>
#<User:0x7f7af20b1eb0 @a=5>

#<User:0x7f7af20b1eb0 @a=5>
#<User:0x7f7af20b1eb0 @a=5>
```

( https://carc.in/#/r/bdn2 )

Environment:

```
uname -a
Linux 14e54b9f81ef 5.10.0-7-rt-amd64 #1 SMP PREEMPT_RT Debian 5.10.40-1 (2021-05-28) x86_64 Linux

---

crystal -v
Crystal 1.1.0-dev [39e5efc] (2021-06-18)

LLVM: 11.1.0
Default target: x86_64-unknown-linux-musl
```

Tested agains different versions on carc (1.0.0, 0.36.1, 0.35.0 etc.), result is the same.

**Comment by straight-shoota** [comment]
`pointerof(value)` returns the pointer of a stack-allocated local variable. That pointer becomes invalid when the scope of this variable is left (which is the case immediately, because it's only available in the block scope).

**Comment by SlayerShadow** [comment]
@straight-shoota, got it. But it should point to specific memory address _with valid objects_ anyway, isn't it? Maybe it is hard to follow StaticArray's pointers but User in example is the `Reference`.

The main difference I see is Value does not have an `object_id`. So I tried pointer-to-pointer combination that probably should point to _some_ objects in memory (at least I expected for that). But I didn't expect to stuck on memory access error.

---

Now I tested with Value only. Works perfect.

```crystal
class User
  def initialize
    @a = 5
  end
end

def show( pointers : Array( Pointer( Void* ) ) )
  p pointers.first.as( Pointer( StaticArray( User?, 2 ) ) ).value
  p pointers.last.as( Pointer( StaticArray( User?, 2 ) ) ).value
end

user = User.new
array = StaticArray[ user, nil ]

raw_array = [
  Pointer( Void* ).new( pointerof( array ).address ),
  Pointer( Void* ).new( pointerof( array ).address )
]

map_array = [ array, array ].map{ |value| Pointer( Void* ).new( pointerof( value ).address ) }

p raw_array, map_array

show raw_array
show map_array
```

Output:

```
[Pointer(Pointer(Void))@0x7ffe59196050, Pointer(Pointer(Void))@0x7ffe59196050]
[Pointer(Pointer(Void))@0x7ffe59195fa0, Pointer(Pointer(Void))@0x7ffe59195fa0]

StaticArray[#<User:0x7fe881ba7eb0 @a=5>, nil]
StaticArray[#<User:0x7fe881ba7eb0 @a=5>, nil]

StaticArray[#<User:0x7fe881ba7eb0 @a=5>, nil]
StaticArray[#<User:0x7fe881ba7eb0 @a=5>, nil]
```

( https://carc.in/#/r/bdn8 )

**Comment by straight-shoota** [comment]
No. It does not point to a valid object. After the block of ˋArray#mapˋ has executed, the pointer is just garbage. If it happens to work for some reason, that's just coincidence. You must not trust that behaviour.

**Comment by SlayerShadow** [comment]
@straight-shoota, thank you! It is indeed works when array consists of `References` only, even if array is union.

I close the issue because it seems not working only when mixing Value and Reference in the same array.
